### ISSUE TITLE
Feature request: HoverArea component

### ISSUE THREAD

**Issue by keyworks** [pinterest/gestalt#2559]
I've implemented hover area behaviour multiple times for the Box component. I wish gestalt had something out of the box. Hover area behaviour isn't always easy to get right, especially w.r.t accessibility so it would be better if gestalt handled it so we don't have bad hover area implementations floating around. 

This is different from TapArea. I just want a component that tells me whether the mouse cursor is over something so that I can show some floating UI. For example, this is the UI I implemented in PDocs to add a expand/closeup button on the top right that only appears after I hover over a diagram:

**Not hovered:**
![Screen Shot 2022-12-10 at 3 18 43 PM](https://user-images.githubusercontent.com/590451/206879088-1ffdc197-5a25-4bf4-a80e-0f7d604a1867.png)

**Hovered:**
![Screen Shot 2022-12-10 at 3 18 37 PM](https://user-images.githubusercontent.com/590451/206879093-471ccb45-e5fa-4c33-b93f-63ea2395bca0.png)

I implemented something like this to achieve this: 
```tsx
import { Box } from "gestalt";
import React, { ComponentProps, useCallback } from "react";

type Props = ComponentProps<typeof Box> & {
  onMouseHover: (hover: boolean) => void;
};

export const HoverArea: React.FC<Props> = (props) => {
  const { onMouseHover, ...boxProps } = props;
  const onMouseHoverOn = useCallback(() => onMouseHover(true), [onMouseHover]);
  const onMouseHoverOff = useCallback(() => onMouseHover(false), [onMouseHover]);
  return (
    <Box
      {...boxProps}
      onMouseOver={onMouseHoverOn}
      onMouseEnter={onMouseHoverOn}
      onMouseLeave={onMouseHoverOff}
    >
      {props.children}
    </Box>
  );
};
```

There are lots of examples out in the wild of people doing this:

- ProductGroupCard: https://phabricator.pinadmin.com/diffusion/P/browse/master/webapp/app/analytics/react/components/ProductGroupCard/ProductGroupCard.js$94-103
- FloatingActionButton: https://phabricator.pinadmin.com/diffusion/P/browse/master/webapp/app/www-unified/components/save/board/BoardOrganization/FloatingActionButton.js$46-54
- UserSettingsPage: https://phabricator.pinadmin.com/diffusion/P/browse/master/webapp/app/www-unified/components/settings/UserSettingsPage/MenuList.js$69-72
- MermaidDiagram: https://phabricator.pinadmin.com/source/bextools/browse/master/common/pdocs/client/components/MermaidDiagram.tsx$80

**Comment by keyworks** [comment]
Yeah I imagine focus/blur is same as hover/leave so that I can tab to activate the hover state. 

I didn't implement it in my case since I didn't think about that accessibility scenario and my case is internal only.

**Comment by dangerismycat** [comment]
Makes sense. I think the main accessibility concern would be ensuring that it supports focus/blur (not a problem) and that there's a visual focus state (possibly a problem, depending on the need). With the implementations you've created, would a blue focus ring be a deal breaker?

**Comment by keyworks** [comment]
I don't think it would be deal breaker for me if the blue focus ring appears. I'm assuming it would only appear during keyboard tab navigation.

**Comment by dangerismycat** [comment]
Sorry for the delay — we discussed this in our team eng meeting today. Given that TapArea's `onTap` is optional, and that it has the mouse/focus handlers you need, it would seem that component would be the best fit here. That also has other built-in niceties like the ability to change the cursor and to connect the hoverable area to the popup content (`accessibilityHaspopup` and `accessibilityControls`). So your example could be something like
```js
<Box>
  {isHovered && (
    <Box id="chart-expand-button">
      <ExpandButton />
    </Box>
  )}

  <TapArea
    accessibilityControls="chart-expand-button"
    accessibilityHaspopup={isHovered}
    onBlur={() => toggleHover(false)}
    onFocus={() => toggleHover(true)}
    onMouseEnter={() => toggleHover(true)}
    onMouseLeave={() => toggleHover(false)}
  >
    <Flowchart />
  </TapArea>
</Box>
```
That's roughly similar to what your HoverArea component is doing, with added accessibility features.

Would that work for this need?

**Comment by keyworks** [comment]
Yeah that works for me.  But the issue for me isn't that I can't use TapArea, but rather that I don't want to re-implement this whenever I need a HoverArea. 

I thought it would be a common enough use case that would warrant its own component. There are some implementation examples of HoverArea-like behaviour that are poorly implemented in pinboard. For example, missing blur/focus handlers or missing mouseOver handler (without it, I found that sometimes hover state doesn't activate).  Unfortunately some of the links I've shared are broken since it looks like things got refactored.
